File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Controllers/CustomerController.cs
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;
using System.Threading.Tasks;
using FluentAssertions;
using RekindleTaf.Configuration;
using RekindleTaf.Configuration.Rest;
using RekindleTaf.Model;
using RestSharp;

namespace RekindleTaf.Controllers;

public class CustomerController(IEnvConfig envConfig, IRestConfig restConfig) : ICustomerController
{
    public async Task<RestResponse> Get(string param = "")
    {
        var request = new RestRequest(envConfig.CustomerEndpoints().CustomerById)
            .AddUrlSegment("customerId", param);
        return await restConfig.RekindleClient().ExecuteGetAsync(request);
    }

    public Customer GetSuccessfully(string param = "")
    {
        var response = Get(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.OK);
        return restConfig.Deserialize<Customer>(response);
    }

    public async Task<RestResponse> GetAll(string param = "")
    {
        var request = new RestRequest(envConfig.CustomerEndpoints().Customers);
        return await restConfig.RekindleClient().ExecuteGetAsync(request);
    }

    public List<Customer> GetAllSuccessfully(string param = "")
    {
        var response = GetAll(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.OK);
        return restConfig.Deserialize<List<Customer>>(response);
    }

    public async Task<RestResponse> Post(Customer param)
    {
        var request = new RestRequest(envConfig.CustomerEndpoints().Customers, Method.Post)
            .AddJsonBody(param);
        return await restConfig.RekindleClient().ExecutePostAsync(request);
    }

    public Customer PostSuccessfully(Customer param)
    {
        var response = Post(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.Created);
        var responseData = restConfig.Deserialize<JsonNode>(response);
        param.Id = responseData["customerId"]!.ToString();
        return param;
    }

    public async Task<RestResponse> Put(Customer param)
    {
        var request = new RestRequest(envConfig.CustomerEndpoints().CustomerById, Method.Put)
            .AddUrlSegment("customerId", param.Id)
            .AddJsonBody(param);
        return await restConfig.RekindleClient().ExecutePutAsync(request);
    }

    public void PutSuccessfully(Customer param)
    {
        var response = Put(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NoContent);
    }

    public async Task<RestResponse> Delete(Customer param)
    {
        var request = new RestRequest(envConfig.CustomerEndpoints().CustomerById, Method.Delete)
            .AddUrlSegment("customerId", param.Id);
        return await restConfig.RekindleClient().ExecuteDeleteAsync(request);
    }

    public void DeleteSuccessfully(Customer param)
    {
        var response = Delete(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.OK);
    }
}

File: typescript-project/rekindle-book-store-aqa/src/rest/controllers/customer-controller.ts
import {Customer} from "@/model/domain-model";
import {AxiosInstance, AxiosResponse, HttpStatusCode} from "axios";
import {CustomerEndpoints} from "@/rest/Endpoint";
import {CustomerCrudController} from "@/rest/controllers/rest-controllers";
import {expect} from "@jest/globals";
import {RestClient} from "@/rest/rekindle-rest-client";
import {inject, injectable} from "tsyringe";

@injectable()
export class CustomerController implements CustomerCrudController {

    private readonly client: AxiosInstance;

    constructor(@inject('RestClient') client: RestClient) {
        this.client = client.getAxios();
    }

    async get(param?: string): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'GET',
            url: CustomerEndpoints.customerById({customerId: param})
        });
    }

    async getSuccessfully(param?: string): Promise<Customer> {
        try {
            const resp = await this.get(param);
            expect(resp.status).toBe(HttpStatusCode.Ok);
            return resp.data;
        } catch (err) {
            throw err;
        }
    }

    async getAll(param?: string): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'GET',
            url: CustomerEndpoints.customers
        });
    }

    async getAllSuccessfully(param?: string): Promise<Customer[]> {
        try {
            const resp = await this.getAll(param);
            expect(resp.status).toBe(HttpStatusCode.Ok);
            return resp.data
        } catch (err) {
            throw err;
        }
    }

    async post(param: Customer): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'POST',
            url: CustomerEndpoints.customers,
            data: param
        });
    }

    async postSuccessfully(param: Customer): Promise<Customer> {
        try {
            let bookstore = param;
            const resp = await this.post(param);
            expect(resp.status).toBe(HttpStatusCode.Created);
            bookstore.id = resp.data.customerId;
            return bookstore;
        } catch (err) {
            throw err;
        }
    }

    async put(param: Customer): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'PUT',
            url: CustomerEndpoints.customerById({customerId: param.id}),
            data: param
        });
    }

    async putSuccessfully(param: Customer): Promise<void> {
        try {
            const resp = await this.put(param);
            expect(resp.status).toBe(HttpStatusCode.NoContent);
        } catch (err) {
            throw err;
        }
    }

    async delete(param: Customer): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'DELETE',
            url: CustomerEndpoints.customerById({customerId: param.id})
        });
    }

    async deleteSuccessfully(param: Customer): Promise<void> {
        try {
            const resp = await this.delete(param);
            expect(resp.status).toBe(HttpStatusCode.Ok);
        } catch (err) {
            throw err;
        }
    }
}